backend/dal/Services/Concrete/ProjectReportService.cs

```diff
@@ -0,0 +1,258 @@
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var snapshots = this.Context.ProjectSnapshots
+                .Include(s => s.Project).ThenInclude(p => p.Agency)
+                .Include(s => s.Project).ThenInclude(p => p.Status)
+                .Where(s => s.SnapshotOn == report.From || s.SnapshotOn == report.To)
+                .ToArray();
+
+            foreach(ProjectSnapshot snapshot in snapshots)
+            {
+                var previousSnapshot = snapshots.Where(s => s.ProjectId == snapshot.ProjectId && s.Id != snapshot.Id).FirstOrDefault();
+                snapshot.BaselineIntegrity = (snapshot?.NetProceeds ?? 0) - (previousSnapshot?.NetProceeds ?? 0);
+            }
+
+            return snapshots;
+        }
+
+        /// <summary>
+        /// Get all project snapshots matching either the 'To' or 'From' dates of the passed report.
+        /// </summary>
+        /// <param name="reportId"></param>
+        /// <exception cref="ArgumentNullException">Argument 'reportId' is required.</exception>
+        /// <exception cref="KeyNotFoundException">Project report does not exist in the datasource.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to view snapshots.</exception>
+        /// <returns></returns>
+        public IEnumerable<ProjectSnapshot> GetSnapshots(int reportId)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var report = this.Context.ProjectReports
+                .FirstOrDefault(p => p.Id == reportId) ?? throw new KeyNotFoundException();
+
+            return GetSnapshots(report);
+        }
+
+        /// <summary>
+        /// Add the specified report to the datasource.
+        /// The 'From' date is set to the 'To' date of the most recent report, if one exists.
+        /// For every SPL project in the system, a new Snapshot is generated using current project data and a 'snapshotOn' time equal to the newly created reports 'To' date.
+        /// </summary>
+        /// <param name="project"></param>
+        /// <exception cref="ArgumentNullException">Argument 'report' is required.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to add report.</exception>
+        /// <returns></returns>
+        public ProjectReport Add(ProjectReport report)
+        {
+            if (report.To == null)
+            {
+                report.To = DateTime.UtcNow;
+            }
+            var mostRecentReport = this.Context.ProjectReports
+                .Where(r => r.ReportTypeId == ReportTypes.SPL)
+                .OrderByDescending(r => r.To)
+                .FirstOrDefault();
+            report.From = mostRecentReport?.To;
+            this.Context.Add(report);
+
+            this.Context.ProjectSnapshots.AddRange(GenerateSnapshots((DateTime)report.To));
```

Comment:
It should be noted that this action will likely result in different report than what the user is viewing presently. This means that you will need to ensure the returned report contains the latest snapshot and the UI is updated. This is particular important before they export the results.

Reply:
You will also need to calculate variance.

Reply:
variance is calculated on all of the get snapshot apis, so I don't think I need to add any additional variance calculations.

Reply:
And yes, since the report and the snapshot data isn't tightly coupled I re-request the snapshot data whenever the report changes.